pkg/arvo/sys/hoon.hoon

```diff
@@ -11484,11 +11485,33 @@
 ::  +swat: deferred +slap
 ::
 ++  swat
-  |=  [tap=(trap vase) gen=hoon]
-  ^-  (trap vase)
+  |=  [tap=ming gen=hoon]
+  ^-  ming
   =/  gun  (~(mint ut p:$:tap) %noun gen)
+  =>  [tap=tap gun=gun]
   |.  ~+
   [p.gun .*(q:$:tap q.gun)]
+::  +shed: deferred +slop
+::
+++  shed
+  |=  [hed=ming tal=ming]
+  ^-  ming
+  =/  cel  [%cell p:$:hed p:$:tal]
+  =>  [cel=cel hed=hed tal=tal]
+  |.  ~+
+  [cel [q:$:hed q:$:tal]]
+::  +shut: deferred +slam
+::
+++  shut
+  |=  [gat=ming sam=ming]
+  ^-  ming
+  =/  typ=type  [%cell p:$:gat p:$:sam]
```

Comment:
gat and sam get fired twice.  I think it would be better to fire them once and store their results in local variables, in case firing is slow for some reason (such as dynamically calculated bunts, which happen sometimes).

Reply:
I think you're stuck with two fires in that you want to defer the Nock evaluation itself.  The first fires have to be done to get the subject type for +mint, which is fine, but you don't want to use the Nock that's produced here, as it would be a kind of early evaluation just "packaged up" in the constructed trap.  Firing again defers the whole Nock computation until it's where it's needed.

(My understanding here comes from how +swat works, but could stand a sanity check.)